src/DI/ContainerBuilder.php

```diff
@@ -463,9 +468,14 @@ private function resolveEntityClass($entity, $recursive = [])
 				throw new ServiceCreationException(sprintf("Factory '%s' used in service '%s' is not callable.", Nette\Utils\Callback::toString($entity), $name[0]));
 			}
 
-			$class = preg_replace('#[|\s].*#', '', PhpReflection::parseAnnotation($reflection, 'return'));
-			if ($class) {
-				$class = $refClass ? PhpReflection::expandClassName($class, $refClass) : ltrim($class, '\\');
+			if (PHP_VERSION_ID >= 70000 && $reflection->hasReturnType()) {
+				$type = $reflection->getReturnType();
+				$class = !$type->isBuiltIn() ? (string) $type : null;
```

Comment:
So you'd let it resolve builtin type correctly and then let `resolveServiceClass()` fail?

Isn't making the `resolveEntityClass()` itself throw exception (ie. `Nette\DI\ServiceCreationException: Type 'string' can not be used as return type in factory 'A\Factory::createString'`) a better solution?